### ISSUE TITLE
Investigate CUDA performance (for LSTMs)

### ISSUE THREAD

**Issue by reidswanson** [deeplearning4j/deeplearning4j#3094]
I've been working on a text prediction problem using a one-hot encoding scheme. I was noticing some strange performance issues while tuning my parameters and came across [this website](https://svail.github.io/rnn_perf/), which suggests cuBLAS is highly sensitive to the minibatch size. The site suggests that [neon's](https://github.com/NervanaSystems/neon) GPU backend is much more stable in this regard. Colleagues have also mentioned that for certain problems it is much faster than most of the other frameworks. So, I thought I'd give it a try.

I reimplemented my training code and did some preliminary testing that seemed to confirm the efficiency strengths of neon. However, I need wildly different hyperparameters from what I need in dl4j (and from what seems to be suggested in the literature) to get similar results. The training in neon also seems much more sensitive to the hyperparameters in terms of making progress on training/validation loss. So, either there's a bug in my code (likely -- the documentation is not very good imo), neon is taking some shortcuts to achieve such high efficiency and/or some other misunderstanding.

I've attached an [eclipse project](https://github.com/deeplearning4j/deeplearning4j/files/862938/lstm-benchmark.tar.gz) that has the core implementations of each. A shaded jar for the dl4j portion can be built with maven to run the associated benchmark script in the `scripts` directory.

## DL4J
The dl4j main program is `test.Benchmark` and expects 3 command line arguments. 1) The batch size 2) the desired sequence length and 3) the dataset file to use. The number of hidden layers and hidden units are baked in as compile time variables. I've got sort of quirky data format I'm using and have provided a main class to generate a dummy dataset and a custom iterator to read it. The data is a sequence of integers. The first indicates the vocabulary size. Then each document is written, first with an int specifying how many tokens are in the document, and then that many ints each representing a token. Finally, a 0 is written when there are no more documents.

To run either of the benchmark programs you need to create a dataset using `test.WriteData.` It expects 2 command line arguments. 1) the fictional vocabulary size and 2) the output file. The number of documents and tokens per document are compile time variables.

Once the dataset is built there's a bash script in the `scripts` folder that will run the program for different batch sizes and output some info to stdout.

## Neon
The neon script is in the `python` folder. It expects several command line options to be set.
- `--data-file` - the path to the file created by `test.WriteData`
- `--limit` - how many tokens to read from the data file. I.e., some number less than the total written but greater than will actually be read, e.g., 50-100k. (Doing it this way simplified this benchmark code)
- `--hidden-units` - how many hidden units in each hidden layer
- `--num-layers` - how many hidden layers to use
- `--time-steps` - the desired number of timesteps

The script will produce output of the same format as the dl4j program.
`toolkit layers units batch_size total_train_time seqs_per_second`
There is also a bash script in the `scripts` directory that will run the python program over various batch sizes.

The comparison isn't exactly apples to apples but should be relatively close. E.g., there isn't an analogous method for manually training one batch at a time like in dl4j. The benchmark method provided in neon is a bit more fine grained but seems fairly close to the training by batches in my dl4j java code.

## Some Observations
Neon isn't exactly stable like I suggested on gitter (and dl4j looks more stable in some ways), but it is quite a bit faster. See the charts below for 2 different sized models.
![lstm-training-neon-vs-dl4j](https://cloud.githubusercontent.com/assets/802748/24221552/fa4184aa-0f0b-11e7-98e8-2d96178f0b6f.png)

There also appears to be a point at which the speed of dl4j drops off and then never recovers from, while this doesn't seem to be the case with neon. Although neon gets less and less stable as the batch size increases.

![lstm-training-neon-vs-dl4j-2](https://cloud.githubusercontent.com/assets/802748/24227788/131c5aec-0f2d-11e7-98fe-1e382479340e.png)
I stopped the dl4j benchmark early due to time constraints, but the trend should be clear.

[lstm-benchmark.tar.gz](https://github.com/deeplearning4j/deeplearning4j/files/862938/lstm-benchmark.tar.gz)

**Comment by AlexDBlack** [comment]
Thanks for the detailed comparison and code.

I wanted to briefly mention that we're currently working on some memory management changes that should provide a boost to performance (for all nets). We'd definitely want to re-run this after those changes have been made.

Also: we don't currently support CuDNN LSTMs, but adding this would give us another performance boost. It would be interesting to compare that to Neon's backend too (they were reporting better performance than CuDNN on CNNs last I checked).

BTW, what's the hardware that those tests were run on?

**Comment by reidswanson** [comment]
No problem, thanks for the awesome toolkit :)

The cpu is an i7 4770k and the gpu is a gtx 1080. I should also mention the vocabulary size in the graphs is 10,000.

**Comment by birinhos** [comment]
I have also test using the same data and the same NN on encog and DL4J. Encog is faster processing epochs (in dl4j "iterations")with  same method, backprop. With encog all cores goes to 97-99% and the fans really start work. DL4J nothing happens cpu at 10-15%.

**Comment by rkfg** [comment]
Here's my comparison with Keras+TensorFlow on the same network architecture, hyperparameters and dataset:
[Keras sample](https://gist.github.com/rkfg/0aa8de658b662e8968993e6a862da808)
[DL4J sample](https://gist.github.com/rkfg/3d356d83668023f2c77cb5ad20cd5497)

The performance difference is quite sad. Keras processes the whole epoch (69368 samples) in 32 seconds on my 1070 GTX, DL4J needs about 288 seconds for the epoch (there's 77076 samples in the set). The size difference is caused by the 90/10 train/test split for Keras which I didn't do for DL4J. Overall, that makes (69368 / 77076) * 288 secods == 259 second for one epoch (of Keras size) on DL4J which is roughly 8 times slower.

The network tries to predict whether a line was said by a specific person or somebody else, hence two classes. The dataset consists of lines like these:
```
someone ¦ hello everyone
txb ¦ hello
```
The network is fed the line char by char, the dictionary and vectorization (or rather token replace) is performed before the training. Presave doesn't help much because the iteration time is vastly bigger than any buffer preparation. Also, the iterator is asynchronous so that shouldn't be an issue in the first place.

**Comment by raver119** [comment]
Performance here should be better now, since workspaces were merged quite a while ago...

**Comment by AlexDBlack** [comment]
Yeah, I'm going to close this - between workspaces and the (just merged) CuDNN LSTM implementation - performance has been considerably improved.

https://github.com/deeplearning4j/deeplearning4j/pull/3339

Simple test on character example (Titan X, Pascal)
- GravesLSTM + workspace single: 356,512ms
- CuDNN LSTM: 72,976ms
